### PR TITLE
Optimize exp2 with batching bit checks

### PR DESCRIPTION
**Pull request by k06a**
I didn't find an easy way to build a gas report.

### CODE REVIEW

**Comment by paulrberg** [comment]
Ooh this is cool, thanks a lot for making this PR! I will review later.

Re gas reports - yeah, a discussed in https://github.com/paulrberg/prb-math/issues/70, the current solution is to do what I did in the [test/gas-estimations](https://github.com/paulrberg/prb-math/tree/test/gas-estimations) branch - add a bunch of `console.log` statements, run each function test set individually, and parse that manually in a Markdown file.

The crux of the problem is that `hardhat-gas-reporter` does not work for internal methods 😬

**Comment by k06a** [comment]
@paulrberg you could make transactions to view methods to make gas reporter work properly, but you will need to subtract calldata cost and `21000`.

**Comment by paulrberg** [comment]
I tested this with the following functions:

```solidity
function measure_exp2(uint256 x) external pure returns (uint256 result) {
    result = x.exp2();
}

function measure_exp2_v2(uint256 x) external pure returns (uint256 result) {
    result = x.exp2_v2();
}
```

Where `exp_v2` is the exp2 function as modified by this PR. Preliminary gas results:

```text
input: 0x30000000000000000
exp2:       23196
exp2_v2: 21762
```

In this case, `v2` was **cheaper** by 1434 gas.

```text
input: 0x4e4acd9e83e425aee6
exp2:       23754
exp2_v2: 23973
```

In this case, `v2` was **more expensive** by 219 gas.

I thus think the best way to decide if this PR should be merged or not is to rebase the [test/gas-estimations](https://github.com/paulrberg/prb-math/tree/test/gas-estimations) branch to use the latest implementations from `main`, push the work on this PR onto that branch, and see how `exp_v2` compares to `exp` across all test inputs.

**Comment by k06a** [comment]
@paulrberg checking to `& 0xFF00` is equal to 8 checks:
- `& 0x8000`
- `& 0x4000`
- `& 0x2000`
- `& 0x1000`
- `& 0x0800`
- `& 0x0400`
- `& 0x0200`
- `& 0x0100`

**Comment by k06a** [comment]
I suspect rewriting this in solidity assembly could win tens of percentages

**Comment by k06a** [comment]
BTW example `0x4e4acd9e83e425aee6` is the worst case for this optimization, since there are no zero bytes in this number representation.

**Comment by paulrberg** [comment]
> `0x4e4acd9e83e425aee6` is the worst case for this optimization

This is a good point. In this case, it does seem like that this PR's optimization will lead to a gas reduction more often than it will lead to a gas increase. Also when the gas is reduced is reduced more than it is increased when it increases.

**Comment by paulrberg** [comment]
Going to merge this onto `staging` and after a few final gas comparisons, I will include it as part of the next release of PRBMath (which is probably going to be v3.0.0).

Thanks for your contribution, @k06a!

File: contracts/PRBMath.sol [hunk 1]
```diff
@@ -131,197 +131,220 @@ library PRBMath {
...
+                }
+                if (x & 0x200000000 > 0) {
+                    result = (result * 0x10000000162E42FF1) >> 64;
+                }
+                if (x & 0x100000000 > 0) {
+                    result = (result * 0x100000000B17217F8) >> 64;
+                }
+            }
+
+            if (x & 0xFF00000000 > 0) {
```
Comment (k06a) [inline contracts/PRBMath.sol:242]: Here is bug, value should be `0xFF000000` instead of `0xFF00000000`

Comment (PaulRBerg) [reply contracts/PRBMath.sol:242]: yup, @andreivladbrg fixed this via #179.

**Comment by k06a** [comment]
@PaulRBerg I sincerely apologise for the bug, it was made unintentionally

**Comment by PaulRBerg** [comment]
No worries, @k06a! I never assumed otherwise

**Pull request merged**